ss_admin_console/main/default/classes/poc/supplier/DataDefinition.cls

```diff
@@ -29,35 +15,94 @@ public inherited sharing class DataDefinition {
     }
 
     public class DataDefinitionBuilder {
-        public DataDefinitionBuilder(SObjectType objectType) {
+        public SObjectType objectType;
+        public List<FieldReference> keyFields = new List<FieldReference>();
+        public List<FieldReference> dataFields = new List<FieldReference>();
+        public List<ReferenceFieldDefinition> referenceFields = new List<ReferenceFieldDefinition>();
 
+        public DataDefinitionBuilder(SObjectType objectType) {
+            this.objectType = objectType;
         }
 
         public DataDefinitionBuilder addKeyField(String keyField) {
+            keyFields.add(FieldReference.build(objectType, keyField));
             return this;
         }
 
         public DataDefinitionBuilder addKeyField(SObjectField keyField) {
+            keyFields.add(FieldReference.build(keyField));
             return this;
         }
 
         public DataDefinitionBuilder addDataField(String dataField) {
+            dataFields.add(FieldReference.build(objectType, dataField));
             return this;
         }
 
         public DataDefinitionBuilder addDataField(SObjectField dataField) {
+            dataFields.add(FieldReference.build(dataField));
             return this;
         }
 
-        public DataDefinitionBuilder addReferenceField(SObjectField referenceField, List<SObjectField> keyFields) {
+        public DataDefinitionBuilder addReferenceField(SObjectField referenceField, List<SObjectField> referencedKeyFields) {
+            ReferenceFieldDefinition rfd = new ReferenceFieldDefinition();
+            rfd.referenceField = referenceField;
+            rfd.referencedKeyFields = toFieldReferences(referencedKeyFields);
+            referenceFields.add(rfd);
+            return this;
+        }
+
+        public DataDefinitionBuilder addReferenceField(SObjectField referenceField, List<String> referencedKeyFields) {
+            ReferenceFieldDefinition rfd = new ReferenceFieldDefinition();
+            if (referenceField != null) {
```

Comment:
First, do all verifications and then do assignments. E.g.:
if (referenceField == null) {
  throw exception;
}
if (referencedKeyField == null || referencedKeyField.isEmpty) {
  throw exception;
}
referenceFields.add(new ReferenceFieldDefinition(referenceField, referencedKeyFields));